cirq/ops/permutation_gate.py

```diff
@@ -0,0 +1,85 @@
+# Copyright 2020 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import Counter
+from typing import Any, Dict, Sequence, Tuple, TYPE_CHECKING
+
+from cirq import protocols, value
+from cirq.ops import raw_types
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality
+class QubitPermutationGate(raw_types.Gate):
+    """A qubit permutation gate specified by a permutation list."""
+
+    def __init__(self, permutation: Sequence[int]):
+        """
+        Args:
+            permutation: A shuffled sequence of integers from 0 to
+                len(permutation) - 1. The entry at offset `i` is the result
+                of permuting `i`.
+        """
+        if not permutation:
+            raise ValueError(f"Invalid permutation (empty): {permutation}")
+
+        if len(Counter(permutation)) < len(permutation):
+            raise ValueError(f"Invalid permutation {permutation} "
+                             f"Each index must appear only once.")
+
+        invalid_indices = list(
+            filter(lambda x: x >= len(permutation) or x < 0, permutation))
+        if len(invalid_indices) > 0:
+            raise ValueError(f"All indices have to satisfy "
+                             f"0 < i < {len(permutation)-1}. "
```

Comment:
Inequalities are incorrect, should be: `f"0 <= i < {len(permutation)}. "

Reply:
ack